openmc/checkvalue.py

```diff
@@ -0,0 +1,337 @@
+    # Initialize the tree at the very first item.
+    tree = [value]
+    index = [0]
+
+    # Traverse the tree.
+    while index[0] != len(tree[0]):
+        # If we are done with this level of the tree, go to the next branch on
+        # the level above this one.
+        if index[-1] == len(tree[-1]):
+            del index[-1]
+            del tree[-1]
+            index[-1] += 1
+            continue
+
+        # Get a string representation of the current index in case we raise an
+        # exception.
+        form = '[' + '{:d}, '*(len(index)-1) + '{:d}]'
+        ind_str = form.format(*index)
+
+        # What is the current item we are looking at?
+        current_item = tree[-1][index[-1]]
+
+        # If this item is of the expected type, then we've reached the bottom
+        # level of this branch.
+        if isinstance(current_item, expected_type):
+            # Is this deep enough?
+            if len(tree) < min_depth:
+                msg = 'Error setting "{0}": The item at {1} does not meet the '\
+                      'minimum depth of {2}'.format(name, ind_str, min_depth)
+                raise TypeError(msg)
+
+            # This item is okay.  Move on to the next item.
+            index[-1] += 1
+
+        # If this item is not of the expected type, then it's either an error or
+        # another level of the tree that we need to pursue deeper.
+        else:
+            if isinstance(current_item, Iterable):
+                # The tree goes deeper here, let's explore it.
+                tree.append(current_item)
+                index.append(0)
+
+                # But first, have we exceeded the max depth?
+                if len(tree) > max_depth:
+                    msg = 'Error setting {0}: Found an iterable at {1}, items '\
+                          'in that iterable exceed the maximum depth of {2}' \
+                          .format(name, ind_str, max_depth)
+                    raise TypeError(msg)
+
+            else:
+                # This item is completely unexpected.
+                msg = "Error setting {0}: Items must be of type '{1}', but " \
+                      "item at {2} is of type '{3}'"\
+                      .format(name, expected_type.__name__, ind_str,
+                              type(current_item).__name__)
+                raise TypeError(msg)
+
+
+def check_length(name, value, length_min, length_max=None):
+    """Ensure that a sized object has length within a given range.
+
+    Parameters
+    ----------
+    name : str
+        Description of value being checked
+    value : collections.Sized
+        Object to check length of
+    length_min : int
+        Minimum length of object
+    length_max : int or None, optional
+        Maximum length of object. If None, it is assumed object must be of
+        length length_min.
+
+    """
+
+    if length_max is None:
+        if len(value) != length_min:
+            msg = 'Unable to set "{0}" to "{1}" since it must be of ' \
```

Comment:
this message doesn't make it clear that the minimum length condition is not being met (assuming that its a minimum condition, and not absolute -- see previous comment).

How about, `...since it must be at least of length "{2}"`